tests fail: _pickle.PicklingError: Can't pickle local object <function live_server.<locals>.run

- [x] I have already searched the issue queue for my error message or symptom


## Details about your operating system

Operating system: Alpine Linux
Version/distribution: edge

Version of Onionshare: 2.6.4

## How did you install OnionShare

- From source

## What is the problem

Tests fail (during the downstream package build). Apparently, they're trying to serialise a _function_ using pickle?

Relevant build output:

```
______________________________________________ TestRangeRequests.test_curl _______________________________________________

self = <tests.test_cli_web.TestRangeRequests object at 0x7f748378f950>, temp_dir = <TemporaryDirectory '/tmp/tmp6jo0iw9n'>
tmpdir = local('/tmp/pytest-of-hugo/pytest-7/test_curl0')
common_obj = <onionshare_cli.common.Common object at 0x7f74837bfcb0>

    @pytest.mark.skipif(sys.platform != "linux", reason="requires Linux")
    @check_unsupported("curl", ["--version"])
    def test_curl(self, temp_dir, tmpdir, common_obj):
        web = web_obj(temp_dir, common_obj, "share", 3)
        web.settings.set("share", "autostop_sharing", False)
    
        download = tmpdir.join("download")
    
>       with live_server(web) as url:
             ^^^^^^^^^^^^^^^^

tests/test_cli_web.py:605: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python3.14/contextlib.py:141: in __enter__
    return next(self.gen)
           ^^^^^^^^^^^^^^
tests/test_cli_web.py:424: in live_server
    proc.start()
/usr/lib/python3.14/multiprocessing/process.py:121: in start
    self._popen = self._Popen(self)
                  ^^^^^^^^^^^^^^^^^
/usr/lib/python3.14/multiprocessing/context.py:230: in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib/python3.14/multiprocessing/context.py:306: in _Popen
    return Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^
/usr/lib/python3.14/multiprocessing/popen_forkserver.py:35: in __init__
    super().__init__(process_obj)
/usr/lib/python3.14/multiprocessing/popen_fork.py:20: in __init__
    self._launch(process_obj)
/usr/lib/python3.14/multiprocessing/popen_forkserver.py:47: in _launch
    reduction.dump(process_obj, buf)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

obj = <Process name='Process-1' parent=3914 initial>, file = <_io.BytesIO object at 0x7f74836e17b0>, protocol = None

    def dump(obj, file, protocol=None):
        '''Replacement for pickle.dump() using ForkingPickler.'''
>       ForkingPickler(file, protocol).dump(obj)
E       _pickle.PicklingError: Can't pickle local object <function live_server.<locals>.run at 0x7f74836c1fe0>
E       when serializing dict item '_target'
E       when serializing multiprocessing.context.Process state
E       when serializing multiprocessing.context.Process object

/usr/lib/python3.14/multiprocessing/reduction.py:60: PicklingError
______________________________________________ TestRangeRequests.test_wget _______________________________________________

self = <tests.test_cli_web.TestRangeRequests object at 0x7f748378fb50>, temp_dir = <TemporaryDirectory '/tmp/tmp6jo0iw9n'>
tmpdir = local('/tmp/pytest-of-hugo/pytest-7/test_wget0')
common_obj = <onionshare_cli.common.Common object at 0x7f7483804590>

    @pytest.mark.skipif(sys.platform != "linux", reason="requires Linux")
    @check_unsupported("wget", ["--version"])
    def test_wget(self, temp_dir, tmpdir, common_obj):
        web = web_obj(temp_dir, common_obj, "share", 3)
        web.settings.set("share", "autostop_sharing", False)
    
        # wget needs a file to exist to continue
        download = tmpdir.join("download")
        download.write("x" * 10)
    
>       with live_server(web) as url:
             ^^^^^^^^^^^^^^^^

tests/test_cli_web.py:629: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python3.14/contextlib.py:141: in __enter__
    return next(self.gen)
           ^^^^^^^^^^^^^^
tests/test_cli_web.py:424: in live_server
    proc.start()
/usr/lib/python3.14/multiprocessing/process.py:121: in start
    self._popen = self._Popen(self)
                  ^^^^^^^^^^^^^^^^^
/usr/lib/python3.14/multiprocessing/context.py:230: in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib/python3.14/multiprocessing/context.py:306: in _Popen
    return Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^
/usr/lib/python3.14/multiprocessing/popen_forkserver.py:35: in __init__
    super().__init__(process_obj)
/usr/lib/python3.14/multiprocessing/popen_fork.py:20: in __init__
    self._launch(process_obj)
/usr/lib/python3.14/multiprocessing/popen_forkserver.py:47: in _launch
    reduction.dump(process_obj, buf)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

obj = <Process name='Process-2' parent=3914 initial>, file = <_io.BytesIO object at 0x7f74832fc4f0>, protocol = None

    def dump(obj, file, protocol=None):
        '''Replacement for pickle.dump() using ForkingPickler.'''
>       ForkingPickler(file, protocol).dump(obj)
E       _pickle.PicklingError: Can't pickle local object <function live_server.<locals>.run at 0x7f74834afd70>
E       when serializing dict item '_target'
E       when serializing multiprocessing.context.Process state
E       when serializing multiprocessing.context.Process object

/usr/lib/python3.14/multiprocessing/reduction.py:60: PicklingError
================================================ short test summary info =================================================
FAILED tests/test_cli_web.py::TestRangeRequests::test_curl - _pickle.PicklingError: Can't pickle local object <function live_server.<locals>.run at 0x7f74836c1fe0>
FAILED tests/test_cli_web.py::TestRangeRequests::test_wget - _pickle.PicklingError: Can't pickle local object <function live_server.<locals>.run at 0x7f74834afd70>
======================================== 2 failed, 258 passed, 5 skipped in 0.36s ========================================
``


## Anything else you'd like to tell us?

```
______________________________________________ TestRangeRequests.test_curl _______________________________________________

self = <tests.test_cli_web.TestRangeRequests object at 0x7f2b42c54d50>, temp_dir = <TemporaryDirectory '/tmp/tmp68a64hzh'>
tmpdir = local('/tmp/pytest-of-hugo/pytest-10/test_curl0')
common_obj = <onionshare_cli.common.Common object at 0x7f2b42fcbe70>

    @pytest.mark.skipif(sys.platform != "linux", reason="requires Linux")
    @check_unsupported("curl", ["--version"])
    def test_curl(self, temp_dir, tmpdir, common_obj):
        web = web_obj(temp_dir, common_obj, "share", 3)
        web.settings.set("share", "autostop_sharing", False)
    
        download = tmpdir.join("download")
    
>       with live_server(web) as url:
             ^^^^^^^^^^^^^^^^

tests/test_cli_web.py:605: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python3.14/contextlib.py:141: in __enter__
    return next(self.gen)
           ^^^^^^^^^^^^^^
tests/test_cli_web.py:424: in live_server
    proc.start()
/usr/lib/python3.14/multiprocessing/process.py:121: in start
    self._popen = self._Popen(self)
                  ^^^^^^^^^^^^^^^^^
/usr/lib/python3.14/multiprocessing/context.py:230: in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib/python3.14/multiprocessing/context.py:306: in _Popen
    return Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^
/usr/lib/python3.14/multiprocessing/popen_forkserver.py:35: in __init__
    super().__init__(process_obj)
/usr/lib/python3.14/multiprocessing/popen_fork.py:20: in __init__
    self._launch(process_obj)
/usr/lib/python3.14/multiprocessing/popen_forkserver.py:47: in _launch
    reduction.dump(process_obj, buf)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

obj = <Process name='Process-1' parent=7516 initial>, file = <_io.BytesIO object at 0x7f2b42dcde40>, protocol = None

    def dump(obj, file, protocol=None):
        '''Replacement for pickle.dump() using ForkingPickler.'''
>       ForkingPickler(file, protocol).dump(obj)
E       _pickle.PicklingError: Can't pickle local object <function live_server.<locals>.run at 0x7f2b42ddf690>
E       when serializing dict item '_target'
E       when serializing multiprocessing.context.Process state
E       when serializing multiprocessing.context.Process object

/usr/lib/python3.14/multiprocessing/reduction.py:60: PicklingError
______________________________________________ TestRangeRequests.test_wget _______________________________________________

self = <tests.test_cli_web.TestRangeRequests object at 0x7f2b42c54f50>, temp_dir = <TemporaryDirectory '/tmp/tmp68a64hzh'>
tmpdir = local('/tmp/pytest-of-hugo/pytest-10/test_wget0')
common_obj = <onionshare_cli.common.Common object at 0x7f2b42c30bb0>

    @pytest.mark.skipif(sys.platform != "linux", reason="requires Linux")
    @check_unsupported("wget", ["--version"])
    def test_wget(self, temp_dir, tmpdir, common_obj):
        web = web_obj(temp_dir, common_obj, "share", 3)
        web.settings.set("share", "autostop_sharing", False)
    
        # wget needs a file to exist to continue
        download = tmpdir.join("download")
        download.write("x" * 10)
    
>       with live_server(web) as url:
             ^^^^^^^^^^^^^^^^

tests/test_cli_web.py:629: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python3.14/contextlib.py:141: in __enter__
    return next(self.gen)
           ^^^^^^^^^^^^^^
tests/test_cli_web.py:424: in live_server
    proc.start()
/usr/lib/python3.14/multiprocessing/process.py:121: in start
    self._popen = self._Popen(self)
                  ^^^^^^^^^^^^^^^^^
/usr/lib/python3.14/multiprocessing/context.py:230: in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib/python3.14/multiprocessing/context.py:306: in _Popen
    return Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^
/usr/lib/python3.14/multiprocessing/popen_forkserver.py:35: in __init__
    super().__init__(process_obj)
/usr/lib/python3.14/multiprocessing/popen_fork.py:20: in __init__
    self._launch(process_obj)
/usr/lib/python3.14/multiprocessing/popen_forkserver.py:47: in _launch
    reduction.dump(process_obj, buf)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

obj = <Process name='Process-2' parent=7516 initial>, file = <_io.BytesIO object at 0x7f2b42eb6ca0>, protocol = None

    def dump(obj, file, protocol=None):
        '''Replacement for pickle.dump() using ForkingPickler.'''
>       ForkingPickler(file, protocol).dump(obj)
E       _pickle.PicklingError: Can't pickle local object <function live_server.<locals>.run at 0x7f2b42c86f00>
E       when serializing dict item '_target'
E       when serializing multiprocessing.context.Process state
E       when serializing multiprocessing.context.Process object

/usr/lib/python3.14/multiprocessing/reduction.py:60: PicklingError
================================================ short test summary info =================================================
FAILED tests/test_cli_web.py::TestRangeRequests::test_curl - _pickle.PicklingError: Can't pickle local object <function live_server.<locals>.run at 0x7f2b42ddf690>
FAILED tests/test_cli_web.py::TestRangeRequests::test_wget - _pickle.PicklingError: Can't pickle local object <function live_server.<locals>.run at 0x7f2b42c86f00>
======================================== 2 failed, 258 passed, 5 skipped in 0.37s ========================================
```

--- a/cli/tests/test_cli_web.py
+++ b/cli/tests/test_cli_web.py
@@ -1,7 +1,6 @@
 import os
 import random
 import re
-import socket
 import subprocess
 import time
 import zipfile
@@ -13,10 +12,11 @@ from io import BytesIO
 
 import pytest
 from contextlib import contextmanager
-from multiprocessing import Process
+from threading import Thread
 from urllib.request import urlopen, Request
 from werkzeug.datastructures import Headers
 from werkzeug.exceptions import RequestedRangeNotSatisfiable
+from werkzeug.serving import make_server
 
 from onionshare_cli.common import Common
 from onionshare_cli.web import Web
@@ -412,35 +412,31 @@ def check_unsupported(cmd: str, args: list):
 
 @contextmanager
 def live_server(web):
-    s = socket.socket()
-    s.bind(("localhost", 0))
-    port = s.getsockname()[1]
-    s.close()
-
-    def run():
-        web.app.run(host="127.0.0.1", port=port, debug=False)
-
-    proc = Process(target=run)
-    proc.start()
+    server = make_server("127.0.0.1", 0, web.app)
+    port = server.socket.getsockname()[1]
+    thread = Thread(target=server.serve_forever, daemon=True)
+    thread.start()
 
     url = "http://127.0.0.1:{}".format(port)
     req = Request(url)
 
-    attempts = 20
-    while True:
-        try:
-            urlopen(req)
-            break
-        except Exception:
-            attempts -= 1
-            if attempts > 0:
-                time.sleep(0.5)
-            else:
-                raise
+    try:
+        attempts = 20
+        while True:
+            try:
+                urlopen(req)
+                break
+            except Exception:
+                attempts -= 1
+                if attempts > 0:
+                    time.sleep(0.5)
+                else:
+                    raise
 
-    yield url + "/download"
-
-    proc.terminate()
+        yield url + "/download"
+    finally:
+        server.shutdown()
+        thread.join(timeout=5)
 
 
 class TestRangeRequests:

